src/scenic/core/pruning.py

```diff
@@ -428,3 +579,60 @@ def relativeHeadingRange(
         tPoints.extend((math.pi, -math.pi))
     rhs = [tp - p for tp in tPoints for p in points]  # TODO improve
     return min(rhs), max(rhs)
+
+
+def percentagePruned(base, newBase):
+    if (
+        base.dimensionality
+        and newBase.dimensionality
+        and base.dimensionality == newBase.dimensionality
+    ):
+        ratio = newBase.size / base.size
+        percent = max(0, 100 * (1.0 - ratio))
+        return percent
+
+    return None
+
+
+def checkCyclical(scenario, A, B):
+    """Check for a potential circular dependency
+
+    Returns True if the scenario would have a circular dependency
+    if A depended on B.
+
+    This function runs DFS on the scenario, with the dependency edges
+    reversed for convenience (this should not affect whether or not a
+    cycle exists).
+    """
+    state = collections.defaultdict(lambda: 0)
+
+    def dfs(target):
+        # Check if the target is already completed/in-process
+        if state[target] == 2:
+            return False
+        elif state[target] == 1:
+            return True
+
+        # Set to in-process
+        state[target] = 1
+
+        # Recurse on children
+        deps = conditionedDeps(target)
+
+        if target is A:
+            deps.append(B)
+
+        for child in deps:
+            if dfs(child):
+                return True
+
+        # Set to completed
+        state[target] = 2
+
+        return False
+
+    return dfs(scenario)
```

Comment:
I don't think this will work: `dependencies(scenario)` will be empty, since `Scenario` is not a subclass of `LazilyEvaluable`. Why not do DFS from `B` and see if you ever reach `A`?

I see you have a test that's supposed to involve this check, so either I'm confused or that test doesn't actually require this code to work correctly.